numpy/tests/test_numpy.py

```diff
@@ -1,762 +1,891 @@
+assert len(arr2) == 3
+
+arr3 = np.array([[1, 2], [3, 4]])
+assert len(arr3) == 2
+
+arr4 = np.array([[[[1.5, -1.5, 2.5], [3.5, -3.5, 4.5], [5.5, -5.5, 6.5]]]])
+assert len(arr4) == 1
+
+
+# test array pow
+assert_equal(np.ones([2, 3]) ** 2, np.array([[1., 1., 1.], [1., 1., 1.]]))
+
+a = np.array([[[1, 2], [3, 4]], [[5, 6], [7, 8]]])
+b = np.array([[[2, 2], [2, 2]], [[3, 3], [3, 3]]])
+
+assert np.allclose(a ** b, np.array([[[1, 4], [9, 16]], [[125, 216], [343, 512]]]))
+
+arr1 = np.array([1, 2, 3, 4, 5])
+assert_equal(arr1 ** 2, np.array([1, 4, 9, 16, 25]))
+assert np.allclose(arr1 ** 2.5, np.array([1.0, 5.656854, 15.588457, 32.0, 55.901699]))
+assert_equal(2 ** arr1, np.array([2, 4, 8, 16, 32]))
+assert np.allclose(2.5 ** arr1, np.array([2.5, 6.25, 15.625, 39.0625, 97.65625]))
+
+arr2 = np.array([[1.33, 2.66], [3.99, 5.33]])
+assert np.allclose(arr2 ** 2, np.array([[1.7689, 7.0756], [15.9201, 28.4089]]))
+assert np.allclose(arr2 ** 2.5, np.array([[2.039995, 11.53995437], [31.80037484, 65.58703869]]))
+assert np.allclose(4 ** arr2, np.array([[6.32033049, 39.94657756],
+                                        [252.47557235, 1618.0046067]]))
+assert np.allclose(5.0 ** arr2, np.array([[8.50413422, 72.32029875],
+                                          [615.0215271, 5315.08388464]]))
+
+
+# test array binary
+a = np.array([[1, 0], [0, 1]], dtype=np.bool_)
+'''
+array([[ True, False],
+       [False,  True]])
+'''
+
+assert_equal(a & True, np.array([[1, 0], [0, 1]], dtype=np.bool_))
+assert_equal(a | True, np.array([[1, 1], [1, 1]], dtype=np.bool_))
+assert_equal(a ^ True, np.array([[0, 1], [1, 0]], dtype=np.bool_))
+assert_equal(~a, np.array([[False,  True], [ True, False]]))
+
+
+# test array trigonometry
+arr1 = np.array([np.pi / 6, np.pi / 4, np.pi / 3, np.pi / 2, np.pi])
+assert np.allclose(np.sin(arr1), np.array([0.5, 0.707107, 0.866025, 1.0, 0.0]))
+assert np.allclose(np.cos(arr1), np.array([0.866025, 0.707107, 0.5, 0.0, -1.0]))
+assert np.allclose(np.tan(arr1), np.array([0.57735, 1.0, 1.73205, np.inf, 0.0]))
+
+arr2 = np.array([0.5, 0.707107, 0.866025, 1.0, 0.0])
+assert np.allclose(np.arcsin(arr2), np.array([np.pi / 6, np.pi / 4, np.pi / 3, np.pi / 2, 0.0]))
+assert np.allclose(np.arccos(arr2), np.array([np.pi / 3, np.pi / 4, np.pi / 6, 0.0, np.pi / 2]))
+assert np.allclose(np.arctan(arr2), np.array([0.463648, 0.61548, 0.713724, 0.785398, 0.0]))
+
+
+# test array exponential
+arr1 = np.array([0.0, 1.0, 2.0, 3.0, 4.0])
+
+assert np.allclose(np.exp(arr1), np.array([1.0, 2.718282, 7.389056, 20.085537, 54.598150]))
+assert np.allclose(np.log(arr1), np.array([-np.inf, 0.0, 0.693147, 1.098612, 1.386294]))
+assert np.allclose(np.log2(arr1), np.array([-np.inf, 0.0, 1.0, 1.584963, 2.0]))
+assert np.allclose(np.log10(arr1), np.array([-np.inf, 0.0, 0.30103, 0.477121, 0.60206]))
+
+
+# test array getitem
+arr1 = np.arange(30).reshape([3, 2, 5])
+
+assert_equal(arr1[0], np.array([[0, 1, 2, 3, 4],
+                                [5, 6, 7, 8, 9]]))
```

Comment:
Hi @16bit-ykiko . Tagging you here as so you can investigate these exact lines.

This is the script I used for `test_main.cpp` to build and run this file -
```
  #include <pybind11/embed.h>
  #include <fstream>
  #include <sstream>
  #include <string>
  
  namespace py = pybind11;
  using namespace pybind11;
  
  int main() {
      py::scoped_interpreter guard{};
      std::ifstream file("test_numpy.py");
      std::stringstream buffer;
      buffer << file.rdbuf();
      std::string script = buffer.str();
      py::exec(script);
  
      return 0;
  }
  ```

Reply:
Kindly try this once and let me know if there are any issues.